File: project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/ast/tuple/Tuple3.java
package org.babyfish.jimmer.sql.ast.tuple;

import java.util.Objects;

public class Tuple3<T1, T2, T3> {

    private final T1 _1;

    private final T2 _2;

    private final T3 _3;

    public Tuple3(T1 _1, T2 _2, T3 _3) {
        this._1 = _1;
        this._2 = _2;
        this._3 = _3;
    }

    public T1 _1() {
        return _1;
    }

    public T2 _2() {
        return _2;
    }

    public T3 _3() {
        return _3;
    }

    @Override
    public int hashCode() {
        return Objects.hash(_1, _2, _3);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Tuple3<?, ?, ?> other = (Tuple3<?, ?, ?>) o;
        return Objects.equals(_1, other._1) &&
                Objects.equals(_2, other._2) &&
                Objects.equals(_3, other._3);
    }

    @Override
    public String toString() {
        return "Tuple3{" +
                "_1=" + _1 +
                ", _2=" + _2 +
                ", _3=" + _3 +
                '}';
    }
}

File: project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/ast/tuple/Tuple3.kt
package org.babyfish.jimmer.sql.ast.tuple

import org.babyfish.jimmer.sql.ast.impl.TupleImplementor
import java.util.function.BiFunction

data class Tuple3<T1, T2, T3>(
    val _1: T1,
    val _2: T2,
    val _3: T3
) : TupleImplementor {

    override fun size(): Int = 3

    override operator fun get(index: Int): Any? =
        when (index) {
            0 -> _1
            1 -> _2
            2 -> _3
            else -> throw IllegalArgumentException("Index must between 0 and ${size() - 1}")
        }

    override fun convert(block: BiFunction<Any?, Int, Any?>): TupleImplementor =
        Tuple3(
            block.apply(_1, 0),
            block.apply(_2, 1),
            block.apply(_3, 2)
        )
}